add RTNL version of netclass collector

raptorsun opened the pull request:
This pull request adds a collector replacing the netclass collector using netlink library.
Sometimes collecting netclass metrics from sysfs `/sys/class/net/<iface>/<metric>` is much slower than using the netlink library . This modification aims to provide a faster way to collect these metrics.
It is also possible to include metrics in netdev collector in this collector, since netlink response message to RTM_GETLINK already contains the metrics.

To activate the collector, we need to disable the netclass collector at the same time. Please refer to the arguments below:
```
./node_exporter --web.listen-address=127.0.0.1:9100 --collector.netclass_rtnl --no-collector.netclass
```

Meanwhile, we can ignore some network devices from the collector by using the argument ` --collector.netclass-rtnl.ignored-devices`:
```
./node_exporter --web.listen-address=127.0.0.1:9100 --collector.netclass_rtnl --no-collector.netclass --collector.netclass-rtnl.ignored-devices='^(lo|bond[0-9]+)$'
```

It provides all the metrics netclass collector has except these 2 metrics:

- addr_assign_type
- name_assign_type

There is a second netlink request to ethtool interface to retrieve these 2 metrics: 
- speed
- duplex

SuperQ commented:
This needs a DCO sign-off. You can use `git commit -s --amend` to add it.

SuperQ commented on collector/netclass_linux_rtnl.go:
```diff
@@ -0,0 +1,387 @@
+)
+
+var (
+	netclassRTNLIgnoredDevices = kingpin.Flag("collector.netclass-rtnl.ignored-devices", "Regexp of net devices to ignore for netclass collector.").Default("^$").String()
```
This flag name needs to match the collector name.

Also, the filename of this collector should match. `netclass_rtnl_linux.go`.

```suggestion
	netclassRTNLIgnoredDevices = kingpin.Flag("collector.netclass_rtnl.ignored-devices", "Regexp of net devices to ignore for netclass collector.").Default("^$").String()
```

raptorsun commented:
Thank you for the review, @SuperQ :) 

I have modified the code according to your comment and added a signature to the commit message.
A second commit is added to give this new collector capability to provide all metrics that `netdev` collector provides. We can use the argument ` --collector.netclass_rtnl.with-stats` to activate these metrics, meanwhile, we need to disable `netdev` collector to avoid scraping the same metrics twice.
```
./node_exporter --web.listen-address=127.0.0.1:9100 --collector.netclass_rtnl  --collector.netclass_rtnl.with-stats --no-collector.netclass --no-collector.netdev
```

jsimonetti commented on collector/netclass_rtnl_linux.go:
```diff
@@ -0,0 +1,489 @@
+
+func (c *netClassRTNLCollector) unmarshalBinary(m *LinkMessage, b []byte) error {
+	l := len(b)
+	if l < unix.SizeofIfInfomsg {
```
This length check will fail depending on which kernel is running on the system.
The constant in package `unix` does not take the kernel version into account.

raptorsun replied:
> As the author of the rtnetlink library I had quite some contact with the netlink library author befor. I know of this tool he created (using netlink) that might add the missing information you are looking for. Perhaps it is worth looking into.. https://github.com/mdlayher/ethtool/blob/main/client.go from glancing the code it should at least cover the cases of speed and duplex.
> 
> it uses the ethtool genetlink family.

Thank you very much for the review ! it is great to have the author of rtnetlink helping me :D 
I will try ethtool netlink interface to fetch the missing 4 metrics and update this PR. 
By the way, do you think it useful to add all the fields in the netlink message RTM_NEWLINK back to the structure rtnetlink.LinkAttributes? for example the carrier state fields.

raptorsun replied:
Which version of the kernel has a different value than 16 for `SizeofIfInfomsg` ? 👀
After a quick check in kernel code, it seems the definition of `struct ifinfomsg` stays unchanged in version [3.19](https://github.com/torvalds/linux/blob/v3.19/include/uapi/linux/rtnetlink.h#L446), [4.0](https://github.com/torvalds/linux/blob/v4.0/include/uapi/linux/rtnetlink.h#L453), [5.18](https://github.com/torvalds/linux/blob/v5.18/include/uapi/linux/rtnetlink.h#L557)

jsimonetti replied:
Hmm, I might have spoken out in error. You are right, that one actually never changed. I apologize for the confusion!

raptorsun replied:
no problem

raptorsun commented:
> As the author of the rtnetlink library I had quite some contact with the netlink library author befor. I know of this tool he created (using netlink) that might add the missing information you are looking for. Perhaps it is worth looking into.. https://github.com/mdlayher/ethtool/blob/main/client.go from glancing the code it should at least cover the cases of speed and duplex.
> 
> it uses the ethtool genetlink family.

Thank you very much for the review! It is great to have the author of `rtnetlink` helping me on the PR :D 

I will try ethtool netlink interface to fetch the missing 4 metrics, so that it can fully replace netclass collector.

Do you think it useful to add all the fields in a netlink RTM_NEWLINK message to the structure `rtnetlink.LinkAttribute`? For example, the carrier state fields and physical port fields.

jsimonetti commented:
> Do you think it useful to add all the fields in a netlink RTM_NEWLINK message to the structure `rtnetlink.LinkAttribute`? For example, the carrier state fields and physical port fields.

I am not sure I understand what you mean by that. The carrier state is already available as `rtnetlink.LinkAttribute.OperationalState` and the physical underlying device as `rtnetlink.LinkAttribute.Master` (for link types that have underlying devices).

raptorsun commented:
> > Do you think it useful to add all the fields in a netlink RTM_NEWLINK message to the structure `rtnetlink.LinkAttribute`? For example, the carrier state fields and physical port fields.
> 
> I am not sure I understand what you mean by that. The carrier state is already available as `rtnetlink.LinkAttribute.OperationalState` and the physical underlying device as `rtnetlink.LinkAttribute.Master` (for link types that have underlying devices).
> 
> Edit: to clarify, if you mean I am missing fields in `LinkAttribute` that are added in later kernels, then I would very much like to add those.
> 
> Edit2: reading your commit again I see now what you mean and you have additional fields in the `LinkAttribute` struct over the one in my library. I would very much like to add those to my library as well.

Yes, I mean to add the new fields in `LinkAttribute` in this PR to `rtnetlink.LinkAttribute`, too. 
I will take care of the PR adding new fields to `rtnetlink.LinkAttribute`, saving you some time for helping other people :)

raptorsun commented:
Hello @jsimonetti , here is the PR to `rtnetlink` package adding new attributes with their decoders.
https://github.com/jsimonetti/rtnetlink/pull/153

jsimonetti commented:
@raptorsun Thanks! I merged the PR and cut a new release. Hopefully that will help with this PR aswel.

raptorsun commented:
> @raptorsun Thanks! I merged the PR and cut a new release. Hopefully that will help with this PR aswel and you can remove some of the duplicate code.

Thank you very much @jsimonetti :) I will update this PR using the updated rtnetlink package.

raptorsun commented:
Code has been updated to use ethtool netlink interface to fetch link mode information, so that we can get 2 more metrics `duplex`  and `speed_byte`.

In some Linux distributions the ethtool netlink interface is not available. So the 2 metrics will not be available when the ethtool family is not available in netlink.

@jsimonetti @SuperQ The PR is ready for review again.

jsimonetti commented:
I can't comment on the prometheus specific stuff, but the netlink code looks good to me.

raptorsun commented:
I tested the new collector in a K8s cluster using OVNK as its CNI. 
The difference of scrape time is significant when the cluster has a heavy workload. 

# netclass collector
## overall scrape time
![image](https://user-images.githubusercontent.com/6040960/195710264-6003cc87-5236-42c2-8ced-59356f3fa27f.png)
Scrape time has touched the 10s watermark of timeout multiple times.

## scrape time by collector
![image](https://user-images.githubusercontent.com/6040960/195710290-d9e0d308-2e11-4318-9459-04a1c1c6a1fb.png)
The slowest collectors are: Netdev 5.6s,  Netclass 5.7s, thermal zone 5.4s.

## node exporter CPU usage
![image](https://user-images.githubusercontent.com/6040960/195711910-1541fce5-c652-4ef5-b598-c25a1be887f5.png)
50 to 150 milli cores is used.

# netclass_rtnl collector
## overall scrape time
![image](https://user-images.githubusercontent.com/6040960/195711635-dd3f3a8e-24ac-4e71-8ab4-c3d34dbf7144.png)
Scrape time is under 4s. 

## scrape time by collector
![image](https://user-images.githubusercontent.com/6040960/195711711-eceac61f-4101-439d-bb27-e0315170b0e1.png)
Netdev 3.6s netclass 3.6s cpu 0.3s

## node exporter CPU usage
![image](https://user-images.githubusercontent.com/6040960/195711951-afc40441-1415-4d3c-9a29-3ba8a092708d.png)
4 milli cores is enough.

discordianfish commented on collector/netclass_rtnl_linux.go:
```diff
@@ -0,0 +1,255 @@
+func pushMetricRTNL(ch chan<- prometheus.Metric, subsystem string, name string, value int64, ifaceName string, valueType prometheus.ValueType) {
+	fieldDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystem, name),
+		fmt.Sprintf("%s value of /sys/class/net/<iface>.", name),
```
This need to be updated, right? Currently this func is the same as pushMetric.

discordianfish reviewed (changes requested):
Can we come up with something that mergess pushMetric in netclass_linux.go, pushStats and pushMetricRTNL? If not, I'd give them more descriptive names and make the new ones part of the struct so it doesn't polute the package namespace with something that is only used here.

raptorsun commented:
> Can we come up with something that mergess pushMetric in netclass_linux.go, pushStats and pushMetricRTNL? If not, I'd give them more descriptive names and make the new ones part of the struct so it doesn't polute the package namespace with something that is only used here.

Thank you very much for the review :) 
I will update the code accordingly this week, trying to merge functions pushMetric, pushStats and pushMetricRTNL into a more generic function shared by both collectors, as well as the check on nil attributes.

raptorsun commented:
The code is refactored so that the the `pushMetric` is generic enough to use elsewhere: doing argument checking, converting the value according to its type and accepting any number of labels as long as the `Desc` has these labels defined. 
2 collectors `netclass` and `netclass_rtnl` share the push metrics function, and cache those already created `prometheus.Desc` structures.

discordianfish reviewed (approved):
Not sure about the 'utils.go' name, maybe move it to collector.go but not sure. Leaving that to @SuperQ. Other than this, LGTM - thanks!

raptorsun commented:
> Not sure about the 'utils.go' name, maybe move it to collector.go but not sure. Leaving that to @SuperQ. Other than this, LGTM - thanks!

Thank you for the review :) 

I was thinking whether `collector.go` is a better place to put `pushMetric` function, too. It is related to collectors but does not manipulate collectors directly, so I put it in a separate toolbox file. 

If we will put it into `collector.go`, I'm ready to update the PR.

SuperQ commented on collector/netclass_rtnl_linux.go:
```diff
@@ -0,0 +1,229 @@
+}
+
+func init() {
+	registerCollector("netclass_rtnl", defaultDisabled, NewNetClassRTNLCollector)
```
Rather than make a whole new collector, what if we made this function like the netdev collector. Where we introduced a flag to pick procfs/sysfs vs netlink via a flag (#2509).

This way we can automatically transition users to the new method, while still offering backwards compatibility.

SuperQ reviewed (approved):
I'm going to merge this as-is and refactor the collector.

raptorsun commented:
> I'm going to merge this as-is and refactor the collector.

Thank you @SuperQ  :) 

I am refactoring the netclass collector to use a flag switching to Netlink interface, according to your last message. I will rebase and make a pull request probably tomorrow. I hope it will save you some time.
